### PR TITLE
MWS: fix editing attachment tiddlers

### PR DESCRIPTION
**Pull request by webplusai**
Issue: https://github.com/TiddlyWiki/TiddlyWiki5/issues/8446
Each time the image was broken on the webpage while following the reproduction steps. It was as a result of the attachment_blob value being set to null when the tiddler was saved.

### CODE REVIEW

File: plugins/tiddlywiki/multiwikiserver/modules/store/sql-tiddler-store.js [hunk 1]
```diff
@@ -151,8 +151,9 @@ SqlTiddlerStore.prototype.processIncomingTiddler = function(tiddlerFields) {
 	const attachmentsEnabled = this.adminWiki.getTiddlerText("$:/config/MultiWikiServer/EnableAttachments","yes") === "yes";
 	const contentTypeInfo = $tw.config.contentTypeInfo[tiddlerFields.type || "text/vnd.tiddlywiki"],
 		isBinary = !!contentTypeInfo && contentTypeInfo.encoding === "base64";
-	if(attachmentsEnabled && isBinary && tiddlerFields.text && tiddlerFields.text.length > attachmentSizeLimit) {
-		const attachment_blob = this.attachmentStore.saveAttachment({
+
```
Comment (Jermolene) [inline plugins/tiddlywiki/multiwikiserver/modules/store/sql-tiddler-store.js:154]: We generally avoid blank lines within function definitions unless they are indispensable for readability.

File: plugins/tiddlywiki/multiwikiserver/modules/store/sql-tiddler-store.js [hunk 2]
```diff
@@ -151,8 +151,9 @@ SqlTiddlerStore.prototype.processIncomingTiddler = function(tiddlerFields) {
...
+	if (attachmentsEnabled && isBinary && ((tiddlerFields.text && tiddlerFields.text.length > attachmentSizeLimit) || (!!tiddlerFields?.attachment_blob && tiddlerFields?.attachment_blob?.length < attachmentSizeLimit))) {
```
Comment (Jermolene) [inline plugins/tiddlywiki/multiwikiserver/modules/store/sql-tiddler-store.js:155]: I am afraid that our coding standards are somewhat old fashioned, and are not currently enforced by tooling. There should not be whitespace between `if` and the following parenthesis `(`.

File: plugins/tiddlywiki/multiwikiserver/modules/store/sql-tiddler-store.js [hunk 3]
```diff
@@ -151,8 +151,9 @@ SqlTiddlerStore.prototype.processIncomingTiddler = function(tiddlerFields) {
...
+		const attachment_blob = tiddlerFields?.attachment_blob ?? this.attachmentStore.saveAttachment({
```
Comment (Jermolene) [inline plugins/tiddlywiki/multiwikiserver/modules/store/sql-tiddler-store.js:156]: Another unusual thing about TiddlyWiki is that in the core we target quite old fashioned JavaScript, aiming to hit compatibility with ECMAScript 5.1. That's why we don't use classes etc. The motivation is to allow TiddlyWiki to run on as wide a range of browsers as possible.

MWS is running in a slightly different situation, where there are very few good reasons for people to be running old versions of Node. So, the plan is to formally adopt a later dialect of JS within MWS. We haven't settled on a version, but I am guessing it might be Node.js 18.

That means that I think it's OK to use the optional chaining operator here, but we should be careful of adopting anything much more recent than that.

File: plugins/tiddlywiki/multiwikiserver/modules/store/sql-tiddler-store.js [hunk 4]
```diff
@@ -275,6 +276,10 @@ SqlTiddlerStore.prototype.saveBagTiddlerWithAttachment = function(incomingTiddle
 Returns {tiddler_id:,bag_name:}
 */
 SqlTiddlerStore.prototype.saveRecipeTiddler = function(incomingTiddlerFields,recipe_name) {
+	var tiddlerInfo = this.sqlTiddlerDatabase.getRecipeTiddler(incomingTiddlerFields.title,recipe_name);
+	if (incomingTiddlerFields.type?.includes('image') && tiddlerInfo?.attachment_blob) {
```
Comment (Jermolene) [inline plugins/tiddlywiki/multiwikiserver/modules/store/sql-tiddler-store.js:280]: The check for "image" within the content type would mean that the attachment handling wouldn't work for (say) videos. There's already code further up this file that identifies whether a tiddler qualifies to be handled as an attachment, and so we should generalise that code.

File: plugins/tiddlywiki/multiwikiserver/modules/store/sql-tiddler-store.js [hunk 5]
```diff
@@ -275,6 +276,10 @@ SqlTiddlerStore.prototype.saveBagTiddlerWithAttachment = function(incomingTiddle
...
+		incomingTiddlerFields['attachment_blob'] = tiddlerInfo.attachment_blob
```
Comment (Jermolene) [inline plugins/tiddlywiki/multiwikiserver/modules/store/sql-tiddler-store.js:281]: I don't think this is right. We should not be setting a field called "attachment_blob" on tiddlers. In the APIs the attachment blob is a separate parameter.

**Review by Jermolene — COMMENTED** [review]
Thank you @webplusai for the speedy response. It's great to see that you've been able to identify a fix so swiftly with unfamiliar code.

I have only conducted a brief review so far as I am away from my desk. I will try it out properly when I am back.

File: plugins/tiddlywiki/multiwikiserver/modules/store/sql-tiddler-store.js (hunk 1 above)
Comment (webplusai) [reply plugins/tiddlywiki/multiwikiserver/modules/store/sql-tiddler-store.js:154]: I intentionally put blank line and space, but let me try to follow the convention of the project.

File: plugins/tiddlywiki/multiwikiserver/modules/store/sql-tiddler-store.js [hunk 6]
```diff
@@ -151,12 +151,15 @@ SqlTiddlerStore.prototype.processIncomingTiddler = function(tiddlerFields) {
 	const attachmentsEnabled = this.adminWiki.getTiddlerText("$:/config/MultiWikiServer/EnableAttachments","yes") === "yes";
 	const contentTypeInfo = $tw.config.contentTypeInfo[tiddlerFields.type || "text/vnd.tiddlywiki"],
 		isBinary = !!contentTypeInfo && contentTypeInfo.encoding === "base64";
-	if(attachmentsEnabled && isBinary && tiddlerFields.text && tiddlerFields.text.length > attachmentSizeLimit) {
-		const attachment_blob = this.attachmentStore.saveAttachment({
+	if(attachmentsEnabled && isBinary && ((tiddlerFields.text && tiddlerFields.text.length > attachmentSizeLimit) || (tiddlerFields?.attachment_blob && tiddlerFields?.attachment_blob?.length < attachmentSizeLimit))) {
```
Comment (Jermolene) [inline plugins/tiddlywiki/multiwikiserver/modules/store/sql-tiddler-store.js:154]: The problem here is that "attachment_blob" is not a tiddler field. It is a property that is returned by some of the methods of SqlTiddlerDatabase.

What we need to do here is (a) look the tiddler up in the database to see if it is currently saved as an attachment and (b) check if the incoming tiddler has a _canonical_uri field with the expected value. If both of those are true, then we save the tiddler fields except _canonical_uri, and maintain the same attachment.

File: plugins/tiddlywiki/multiwikiserver/modules/store/sql-tiddler-store.js [hunk 7]
```diff
@@ -143,20 +147,25 @@ SqlTiddlerStore.prototype.processOutgoingTiddler = function(tiddlerFields,tiddle
 
 /*
 */
-SqlTiddlerStore.prototype.processIncomingTiddler = function(tiddlerFields) {
+SqlTiddlerStore.prototype.processIncomingTiddler = function(tiddlerFields, metadata) {
 	let attachmentSizeLimit = $tw.utils.parseNumber(this.adminWiki.getTiddlerText("$:/config/MultiWikiServer/AttachmentSizeLimit"));
 	if(attachmentSizeLimit < 100 * 1024) {
 		attachmentSizeLimit = 100 * 1024;
 	}
+	var tiddlerInfo = this.sqlTiddlerDatabase[this.fetchTiddlerFuncMap[metadata.entityType]](metadata.title,metadata.name);
```
Comment (Jermolene) [inline plugins/tiddlywiki/multiwikiserver/modules/store/sql-tiddler-store.js:155]: I am not keen on the indirection here. I understand the motivation of avoiding code duplication, but the result is almost as complexity as the duplication.

In any case, I think calling getRecipeTiddler or getBagTiddler is overkill here. We only want to know the attachment status of the tiddler, while those methods perform a second query to retrieve the tiddler fields which is not needed in this context.

Therefore I think it would be better to handle things as follows:

* Introduce new database methods getBagTiddlerAttachmentBlob and getRecipeTiddlerAttachmentBlob that just do the lookup of the attachment_blob property
* change processIncomingTiddler to take an additional parameter "attachment_blob" instead of "metadata"
* In the two calling sites of "SqlTiddlerStore.prototype.saveBagTiddler" and "SqlTiddlerStore.prototype.saveBagTiddler", invoke the appropriate one of the new methods to pass the attachment_blob

**Review by Jermolene — COMMENTED** [review]
Thanks @webplusai much appreciated

File: plugins/tiddlywiki/multiwikiserver/modules/store/sql-tiddler-store.js [hunk 8]
```diff
@@ -147,18 +143,17 @@ SqlTiddlerStore.prototype.processOutgoingTiddler = function(tiddlerFields,tiddle
 
 /*
 */
-SqlTiddlerStore.prototype.processIncomingTiddler = function(tiddlerFields, metadata) {
+SqlTiddlerStore.prototype.processIncomingTiddler = function(tiddlerFields, exisiting_attachment_blob) {
```
Comment (Jermolene) [inline plugins/tiddlywiki/multiwikiserver/modules/store/sql-tiddler-store.js:146]: Typo exisiting_attachment_blob --> existing_attachment_blob

File: plugins/tiddlywiki/multiwikiserver/modules/store/sql-tiddler-store.js [hunk 9]
```diff
@@ -143,20 +143,24 @@ SqlTiddlerStore.prototype.processOutgoingTiddler = function(tiddlerFields,tiddle
...
 	let attachmentSizeLimit = $tw.utils.parseNumber(this.adminWiki.getTiddlerText("$:/config/MultiWikiServer/AttachmentSizeLimit"));
 	if(attachmentSizeLimit < 100 * 1024) {
 		attachmentSizeLimit = 100 * 1024;
 	}
 	const attachmentsEnabled = this.adminWiki.getTiddlerText("$:/config/MultiWikiServer/EnableAttachments","yes") === "yes";
 	const contentTypeInfo = $tw.config.contentTypeInfo[tiddlerFields.type || "text/vnd.tiddlywiki"],
 		isBinary = !!contentTypeInfo && contentTypeInfo.encoding === "base64";
-	if(attachmentsEnabled && isBinary && tiddlerFields.text && tiddlerFields.text.length > attachmentSizeLimit) {
-		const attachment_blob = this.attachmentStore.saveAttachment({
+	const shouldProcessAttachment = ((tiddlerFields.text && tiddlerFields.text.length > attachmentSizeLimit) || (exisiting_attachment_blob && exisiting_attachment_blob?.length < attachmentSizeLimit))
```
Comment (Jermolene) [inline plugins/tiddlywiki/multiwikiserver/modules/store/sql-tiddler-store.js:154]: What is the logic for checking the size of the existing attachment blob?

File: plugins/tiddlywiki/multiwikiserver/modules/store/sql-tiddler-store.js [hunk 10]
```diff
@@ -143,20 +143,24 @@ SqlTiddlerStore.prototype.processOutgoingTiddler = function(tiddlerFields,tiddle
...
+	if(attachmentsEnabled && isBinary && shouldProcessAttachment) {
```
Comment (Jermolene) [inline plugins/tiddlywiki/multiwikiserver/modules/store/sql-tiddler-store.js:155]: It looks like the logic here would make it impossible to modify an existing attachment. We only want to skip saving the attachment if we have received a _canonical_uri field with the same value. If the tiddler is coming in as an ordinary tiddlers (without a _canonical_uri field) then we would want to save the new value.

**Review by Jermolene — COMMENTED** [review]
Hi @webplusai thank you, this is looking good. My concerns now are around the edge cases.

Have you seen the test framework in $:/plugins/tiddlywiki/multiwikiserver/commands/mws-test-server.js? I think we need to look at introducing some tests for the attachment handling logic. I would hope it could be done relatively easily if we could figure out a convenient way to bring down the attachment size limit so that the tests can work with attachments of 10s of bytes.

File: editions/test/tiddlers/tests/test-attachment.js [hunk 11]
```diff
@@ -0,0 +1,168 @@
+/*\
+title: test-attachment.js
+type: application/javascript
+tags: [[$:/tags/test-spec]]
```
Comment (Jermolene) [inline editions/test/tiddlers/tests/test-attachment.js:4]: These tests should reside within the multiwikiserver plugin, and the tiddlywiki.info for the `editions/test` edition should be modified to include the multiwikiserver plugin. (This is the approach that the geospatial plugin uses for its tests).

File: plugins/tiddlywiki/multiwikiserver/modules/store/attachments.js [hunk 12]
```diff
@@ -69,19 +69,20 @@ AttachmentStore.prototype.saveAttachment = function(options) {
...
+		type: options.type,
 	},null,4));
 	return contentHash;
 };
 
 /*
 Adopts an attachment file into the store
 */
-AttachmentStore.prototype.adoptAttachment = function(incomingFilepath,type,hash) {
+AttachmentStore.prototype.adoptAttachment = function(incomingFilepath,type,hash,canonicalUri) {
```
Comment (Jermolene) [inline plugins/tiddlywiki/multiwikiserver/modules/store/attachments.js:85]: I think we should consistently use `_canonical_uri` as the variable name

File: plugins/tiddlywiki/multiwikiserver/modules/store/sql-tiddler-store.js [hunk 13]
```diff
@@ -143,30 +143,51 @@ SqlTiddlerStore.prototype.processOutgoingTiddler = function(tiddlerFields,tiddle
...
-			attachment_blob: null
-		};
-	}
+  const attachmentsEnabled = this.adminWiki.getTiddlerText("$:/config/MultiWikiServer/EnableAttachments", "yes") === "yes";
+  const contentTypeInfo = $tw.config.contentTypeInfo[tiddlerFields.type || "text/vnd.tiddlywiki"];
+  const isBinary = !!contentTypeInfo && contentTypeInfo.encoding === "base64";
+
+  let shouldProcessAttachment = tiddlerFields.text && tiddlerFields.text.length <= attachmentSizeLimit;
+
+  if (existing_attachment_blob) {
```
Comment (Jermolene) [inline plugins/tiddlywiki/multiwikiserver/modules/store/sql-tiddler-store.js:157]: Another coding style weirdness is that we don't include whitespace between `if` and the parenthesis.

@pmario may be able to advise on linter settings that can help with TW coding styles

**Review by Jermolene — COMMENTED** [review]
Hi @webplusai this is looking very good. The new tests are very clear and much appreciated.

There are a couple of minor points but otherwise I think this is ready to merge

File: package-lock.json [hunk 14]
```diff
@@ -41,23 +41,24 @@
       }
     },
     "node_modules/@babel/helper-validator-identifier": {
-      "version": "7.22.20",
-      "resolved": "https://registry.npmjs.org/@babel/helper-validator-identifier/-/helper-validator-identifier-7.22.20.tgz",
-      "integrity": "sha512-Y4OZ+ytlatR8AI+8KZfKuL5urKp7qey08ha31L8b3BwewJAoJamTzyvxPR/5D+KkdJCGPq/+8TukHBlY10FX9A==",
+      "version": "7.24.7",
```
Comment (Jermolene) [inline package-lock.json:44]: I don't think these changes to package-lock.json are needed anymore.

File: package.json [hunk 15]
```diff
@@ -31,8 +31,10 @@
     "node": ">=0.8.2"
   },
   "scripts": {
-	"start": "node ./tiddlywiki.js ./editions/multiwikiserver --mws-load-plugin-bags --build load-mws-demo-data --mws-listen",
-    "test": "node ./tiddlywiki.js ./editions/test --verbose --version --build index && node ./tiddlywiki ./editions/multiwikiserver/ --build load-mws-demo-data --mws-listen --mws-test-server http://127.0.0.1:8080/ --quit",
+    "start": "node ./tiddlywiki.js ./editions/multiwikiserver --mws-load-plugin-bags --build load-mws-demo-data --mws-listen",
+    "build:test-edition": "node ./tiddlywiki.js ./editions/test --verbose --version --build index",
+    "build:multiwikiserver-edition": "node ./tiddlywiki.js ./editions/multiwikiserver/ --build load-mws-demo-data --mws-listen --mws-test-server http://127.0.0.1:8080/ --quit",
```
Comment (Jermolene) [inline package.json:36]: It looks like this should perhaps be `test:multiwikiserver-edition`?

File: plugins/tiddlywiki/jasmine/files/jasmine-core/lib/jasmine-core/jasmine.js [hunk 16]
```diff
@@ -170,7 +170,7 @@ getJasmineRequireObj().base = function(j$, jasmineGlobal) {
    * Default number of milliseconds Jasmine will wait for an asynchronous spec to complete.
    * @name jasmine.DEFAULT_TIMEOUT_INTERVAL
    */
-  j$.DEFAULT_TIMEOUT_INTERVAL = 5000;
+  j$.DEFAULT_TIMEOUT_INTERVAL = 30000;
```
Comment (Jermolene) [inline plugins/tiddlywiki/jasmine/files/jasmine-core/lib/jasmine-core/jasmine.js:173]: Files within the `plugins/tiddlywiki/jasmine/files` directory are taken directly from the Jasmine release. Any changes here will be overwritten the next time Jasmine is upgraded. If the increased timeout is necessary we should explore whether it can be configured in the TiddlyWIki JS wrappers

**Review by Jermolene — COMMENTED** [review]
Thank you again for you patience @webplusai. This is excellent quality of work.

There's a couple of minor points below, but otherwise this is ready to merge.

**Comment by Jermolene** [comment]
Thank you @webplusai for your patience, and congratulations on landing a significant PR.

**Comment by Jermolene** [comment]
Hi @webplusai it turns out that there is a small problem with the tests – they fail if the directory `./editions/test/test-store` does not exist. That means that the tests are failing in CI at the moment so it would be great to sort this out as soon as possible.
